File: libraries/react-native-iap/src/__tests__/standardized-offer-types.test.ts
/* eslint-disable @typescript-eslint/no-non-null-assertion */
/**
 * Tests for standardized offer types and the input field naming convention.
 *
 * Key principle tested here:
 * - Response types (DiscountOffer, SubscriptionOffer, etc.) use Android suffix for platform-specific fields
 * - Input types (RequestPurchaseAndroidProps) do NOT use Android suffix since the parent type indicates platform
 *
 * Example:
 * - Response: DiscountOffer.offerTokenAndroid (suffix because it's cross-platform type)
 * - Input: RequestPurchaseAndroidProps.offerToken (no suffix, parent type is Android-specific)
 */

import type {
  DiscountOffer,
  SubscriptionOffer,
  ProductAndroid,
  ProductSubscriptionAndroid,
  RequestPurchaseAndroidProps,
  InstallmentPlanDetailsAndroid,
  PendingPurchaseUpdateAndroid,
  PurchaseAndroid,
} from '../types';

describe('Standardized Offer Types', () => {
  describe('DiscountOffer', () => {
    it('should have correct structure with Android-specific fields', () => {
      const discountOffer: DiscountOffer = {
        id: 'summer_sale_2025',
        displayPrice: '$4.99',
        price: 4.99,
        currency: 'USD',
        type: 'one-time',
        offerTokenAndroid: 'token_abc123',
        offerTagsAndroid: ['summer', 'sale'],
        fullPriceMicrosAndroid: '9990000',
        percentageDiscountAndroid: 50,
        discountAmountMicrosAndroid: '4990000',
        formattedDiscountAmountAndroid: '$5.00 OFF',
        validTimeWindowAndroid: {
          startTimeMillis: '1704067200000',
          endTimeMillis: '1735689599000',
        },
        limitedQuantityInfoAndroid: {
          maximumQuantity: 3,
          remainingQuantity: 2,
        },
      };

      expect(discountOffer.id).toBe('summer_sale_2025');
      expect(discountOffer.displayPrice).toBe('$4.99');
      expect(discountOffer.price).toBe(4.99);
      expect(discountOffer.currency).toBe('USD');
      expect(discountOffer.type).toBe('one-time');
      expect(discountOffer.offerTokenAndroid).toBe('token_abc123');
      expect(discountOffer.offerTagsAndroid).toEqual(['summer', 'sale']);
      expect(discountOffer.fullPriceMicrosAndroid).toBe('9990000');
      expect(discountOffer.percentageDiscountAndroid).toBe(50);
      expect(discountOffer.validTimeWindowAndroid?.startTimeMillis).toBe(
        '1704067200000',
      );
      expect(discountOffer.limitedQuantityInfoAndroid?.maximumQuantity).toBe(3);
    });

    it('should support all DiscountOfferType values', () => {
      const introductory: DiscountOffer = {
        displayPrice: '$0.99',
        price: 0.99,
        currency: 'USD',
        type: 'introductory',
      };

      const promotional: DiscountOffer = {
        displayPrice: '$1.99',
        price: 1.99,
        currency: 'USD',
        type: 'promotional',
      };

      const oneTime: DiscountOffer = {
        displayPrice: '$2.99',
        price: 2.99,
        currency: 'USD',
        type: 'one-time',
      };

      expect(introductory.type).toBe('introductory');
      expect(promotional.type).toBe('promotional');
      expect(oneTime.type).toBe('one-time');
    });
  });

  describe('SubscriptionOffer', () => {
    it('should have correct structure with cross-platform fields', () => {
      const subscriptionOffer: SubscriptionOffer = {
        id: 'intro_monthly',
        displayPrice: '$0.00',
        price: 0,
        currency: 'USD',
        type: 'introductory',
        paymentMode: 'free-trial',
        period: {unit: 'week', value: 1},
        periodCount: 1,
        basePlanIdAndroid: 'monthly_base',
        offerTokenAndroid: 'offer_token_abc',
        offerTagsAndroid: ['trial'],
      };

      expect(subscriptionOffer.id).toBe('intro_monthly');
      expect(subscriptionOffer.displayPrice).toBe('$0.00');
      expect(subscriptionOffer.price).toBe(0);
      expect(subscriptionOffer.paymentMode).toBe('free-trial');
      expect(subscriptionOffer.period?.unit).toBe('week');
      expect(subscriptionOffer.period?.value).toBe(1);
      expect(subscriptionOffer.periodCount).toBe(1);
      expect(subscriptionOffer.basePlanIdAndroid).toBe('monthly_base');
      expect(subscriptionOffer.offerTokenAndroid).toBe('offer_token_abc');
    });

    it('should support iOS-specific fields', () => {
      const iosOffer: SubscriptionOffer = {
        id: 'promo_ios',
        displayPrice: '$4.99',
        price: 4.99,
        type: 'promotional',
        keyIdentifierIOS: 'key_123',
        nonceIOS: 'uuid-nonce-456',
        signatureIOS: 'signature_base64',
        timestampIOS: 1704067200000,
        numberOfPeriodsIOS: 3,
        localizedPriceIOS: '$4.99',
      };

      expect(iosOffer.keyIdentifierIOS).toBe('key_123');
      expect(iosOffer.nonceIOS).toBe('uuid-nonce-456');
      expect(iosOffer.signatureIOS).toBe('signature_base64');
      expect(iosOffer.timestampIOS).toBe(1704067200000);
      expect(iosOffer.numberOfPeriodsIOS).toBe(3);
    });

    it('should support all PaymentMode values', () => {
      const freeTrial: SubscriptionOffer = {
        id: 'trial',
        displayPrice: '$0.00',
        price: 0,
        type: 'introductory',
        paymentMode: 'free-trial',
      };

      const payAsYouGo: SubscriptionOffer = {
        id: 'payg',
        displayPrice: '$2.99',
        price: 2.99,
        type: 'introductory',
        paymentMode: 'pay-as-you-go',
      };

      const payUpFront: SubscriptionOffer = {
        id: 'upfront',
        displayPrice: '$9.99',
        price: 9.99,
        type: 'introductory',
        paymentMode: 'pay-up-front',
      };

      const unknown: SubscriptionOffer = {
        id: 'unknown',
        displayPrice: '$0.00',
        price: 0,
        type: 'introductory',
        paymentMode: 'unknown',
      };

      expect(freeTrial.paymentMode).toBe('free-trial');
      expect(payAsYouGo.paymentMode).toBe('pay-as-you-go');
      expect(payUpFront.paymentMode).toBe('pay-up-front');
      expect(unknown.paymentMode).toBe('unknown');
    });

    it('should support pricingPhasesAndroid', () => {
      const offerWithPhases: SubscriptionOffer = {
        id: 'offer_with_phases',
        displayPrice: '$0.00',
        price: 0,
        type: 'introductory',
        pricingPhasesAndroid: {
          pricingPhaseList: [
            {
              billingCycleCount: 0,
              billingPeriod: 'P1W',
              formattedPrice: '$0.00',
              priceAmountMicros: '0',
              priceCurrencyCode: 'USD',
              recurrenceMode: 3,
            },
            {
              billingCycleCount: 0,
              billingPeriod: 'P1M',
              formattedPrice: '$9.99',
              priceAmountMicros: '9990000',
              priceCurrencyCode: 'USD',
              recurrenceMode: 1,
            },
          ],
        },
      };

      const pricingPhases = offerWithPhases.pricingPhasesAndroid;
      expect(pricingPhases).toBeDefined();
      expect(pricingPhases!.pricingPhaseList).toHaveLength(2);
      expect(pricingPhases!.pricingPhaseList[0]!.billingPeriod).toBe('P1W');
      expect(pricingPhases!.pricingPhaseList[0]!.recurrenceMode).toBe(3);
      expect(pricingPhases!.pricingPhaseList[1]!.formattedPrice).toBe('$9.99');
    });
  });

  describe('Product with offer fields', () => {
    it('should have correct type structure for ProductAndroid with discountOffers', () => {
      // This test validates the type structure rather than the API call
      // The actual fetchProducts conversion is tested in index.test.ts
      const mockProduct: ProductAndroid = {
        id: 'test_product',
        title: 'Test Product',
        description: 'A test product',
        displayName: 'Test',
        displayPrice: '$9.99',
        price: 9.99,
        currency: 'USD',
        platform: 'android',
        type: 'in-app',
        nameAndroid: 'Test Product',
        discountOffers: [
          {
            id: 'discount_001',
            displayPrice: '$4.99',
            price: 4.99,
            currency: 'USD',
            type: 'one-time',
            offerTokenAndroid: 'disc_token',
          },
        ],
      };

      const discountOffers = mockProduct.discountOffers;
      expect(discountOffers).toBeDefined();
      expect(discountOffers).toHaveLength(1);
      expect(discountOffers![0]!.id).toBe('discount_001');
      expect(discountOffers![0]!.offerTokenAndroid).toBe('disc_token');
    });

    it('should have correct type structure for ProductSubscriptionAndroid with subscriptionOffers', () => {
      // This test validates the type structure rather than the API call
      // The actual fetchProducts conversion is tested in index.test.ts
      const mockSubscription: ProductSubscriptionAndroid = {
        id: 'subscription_product',
        title: 'Premium Subscription',
        description: 'Monthly premium subscription',
        displayName: 'Premium',
        displayPrice: '$9.99',
        price: 9.99,
        currency: 'USD',
        platform: 'android',
        type: 'subs',
        nameAndroid: 'Premium Subscription',
        subscriptionOfferDetailsAndroid: [],
        subscriptionOffers: [
          {
            id: 'sub_intro',
            displayPrice: '$0.00',
            price: 0,
            type: 'introductory',
            paymentMode: 'free-trial',
            basePlanIdAndroid: 'monthly',
            offerTokenAndroid: 'sub_token',
          },
        ],
      };

      const subscriptionOffers = mockSubscription.subscriptionOffers;
      expect(subscriptionOffers).toBeDefined();
      expect(subscriptionOffers).toHaveLength(1);
      expect(subscriptionOffers![0]!.id).toBe('sub_intro');
      expect(subscriptionOffers![0]!.paymentMode).toBe('free-trial');
      expect(subscriptionOffers![0]!.basePlanIdAndroid).toBe('monthly');
      expect(subscriptionOffers![0]!.offerTokenAndroid).toBe('sub_token');
    });
  });

  describe('Building purchase request with subscriptionOffers', () => {
    it('should extract offerTokenAndroid for purchase request', () => {
      const subscriptionOffers: SubscriptionOffer[] = [
        {
          id: 'offer1',
          displayPrice: '$0.00',
          price: 0,
          type: 'introductory',
          paymentMode: 'free-trial',
          basePlanIdAndroid: 'monthly',
          offerTokenAndroid: 'token_1',
        },
        {
          id: 'offer2',
          displayPrice: '$4.99',
          price: 4.99,
          type: 'promotional',
          basePlanIdAndroid: 'yearly',
          offerTokenAndroid: 'token_2',
        },
      ];

      // Simulate building purchase request
      const offers = subscriptionOffers
        .filter((offer) => offer.offerTokenAndroid)
        .map((offer) => ({
          sku: 'premium_sub',
          offerToken: offer.offerTokenAndroid!,
        }));

      expect(offers).toHaveLength(2);
      expect(offers[0]).toEqual({sku: 'premium_sub', offerToken: 'token_1'});
      expect(offers[1]).toEqual({sku: 'premium_sub', offerToken: 'token_2'});
    });

    it('should filter out offers without offerTokenAndroid', () => {
      const subscriptionOffers: SubscriptionOffer[] = [
        {
          id: 'ios_offer',
          displayPrice: '$4.99',
          price: 4.99,
          type: 'promotional',
          // No offerTokenAndroid - iOS only
          keyIdentifierIOS: 'key_123',
          signatureIOS: 'sig_abc',
        },
        {
          id: 'android_offer',
          displayPrice: '$0.00',
          price: 0,
          type: 'introductory',
          offerTokenAndroid: 'token_android',
        },
      ];

      const androidOffers = subscriptionOffers
        .filter((offer) => offer.offerTokenAndroid)
        .map((offer) => ({
          sku: 'sub_id',
          offerToken: offer.offerTokenAndroid!,
        }));

      expect(androidOffers).toHaveLength(1);
      expect(androidOffers[0]!.offerToken).toBe('token_android');
    });
  });

  describe('RequestPurchaseAndroidProps with offerToken', () => {
    it('should support offerToken for one-time purchase discounts', () => {
      // This tests the type structure for one-time purchase discount offers
      // introduced in Google Play Billing Library 7.0
      // Note: Input fields no longer have Android suffix (parent type indicates platform)
      const purchaseRequest: RequestPurchaseAndroidProps = {
        skus: ['premium_upgrade'],
        offerToken: 'discount_offer_token_abc123',
        isOfferPersonalized: false,
        obfuscatedAccountId: 'account_123',
        obfuscatedProfileId: 'profile_456',
      };

      expect(purchaseRequest.skus).toEqual(['premium_upgrade']);
      expect(purchaseRequest.offerToken).toBe('discount_offer_token_abc123');
      expect(purchaseRequest.isOfferPersonalized).toBe(false);
      expect(purchaseRequest.obfuscatedAccountId).toBe('account_123');
      expect(purchaseRequest.obfuscatedProfileId).toBe('profile_456');
    });

    it('should allow offerToken to be optional', () => {
      const purchaseRequestWithoutOffer: RequestPurchaseAndroidProps = {
        skus: ['regular_product'],
        // No offerToken - regular purchase without discount
      };

      expect(purchaseRequestWithoutOffer.skus).toEqual(['regular_product']);
      expect(purchaseRequestWithoutOffer).not.toHaveProperty('offerToken');
    });

    it('should extract offerTokenAndroid from DiscountOffer for purchase input', () => {
      // Simulate getting a product with discount offers
      // Note: Response types (DiscountOffer) keep Android suffix
      const discountOffer: DiscountOffer = {
        id: 'flash_sale',
        displayPrice: '$2.99',
        price: 2.99,
        currency: 'USD',
        type: 'one-time',
        offerTokenAndroid: 'flash_sale_token_xyz',
        percentageDiscountAndroid: 50,
      };

      // Build purchase request using the offer token from the discount offer
      // Input field uses offerToken (no suffix), value comes from response's offerTokenAndroid
      const purchaseRequest: RequestPurchaseAndroidProps = {
        skus: ['premium_upgrade'],
        offerToken: discountOffer.offerTokenAndroid ?? undefined,
      };

      expect(purchaseRequest.offerToken).toBe('flash_sale_token_xyz');
      expect(purchaseRequest.offerToken).toBe(discountOffer.offerTokenAndroid);
    });

    it('should support isOfferPersonalized for EU compliance', () => {
      // isOfferPersonalized indicates when the price was customized for this user
      // Required for EU Digital Services Act compliance
      // Note: Input field uses isOfferPersonalized (no Android suffix)
      const personalizedRequest: RequestPurchaseAndroidProps = {
        skus: ['premium_product'],
        isOfferPersonalized: true,
      };

      const nonPersonalizedRequest: RequestPurchaseAndroidProps = {
        skus: ['premium_product'],
        isOfferPersonalized: false,
      };

      expect(personalizedRequest.isOfferPersonalized).toBe(true);
      expect(nonPersonalizedRequest.isOfferPersonalized).toBe(false);
    });

    it('should combine discountOffers offerTokenAndroid with purchase request', () => {
      // Full workflow: product → discount offer → purchase request
      // Response type (ProductAndroid.discountOffers) uses offerTokenAndroid
      // Input type (purchase request) uses offerToken (no suffix)
      const mockProduct: ProductAndroid = {
        id: 'consumable_gems',
        title: '100 Gems',
        description: 'A pack of 100 gems',
        displayName: 'Gems Pack',
        displayPrice: '$4.99',
        price: 4.99,
        currency: 'USD',
        platform: 'android',
        type: 'in-app',
        nameAndroid: '100 Gems',
        discountOffers: [
          {
            id: 'summer_sale',
            displayPrice: '$2.49',
            price: 2.49,
            currency: 'USD',
            type: 'one-time',
            offerTokenAndroid: 'summer_sale_offer_token', // Response field keeps suffix
            percentageDiscountAndroid: 50,
          },
        ],
      };

      // Get the offer token from product's discount offers (response field has suffix)
      const discountOffers = mockProduct.discountOffers;
      expect(discountOffers).toBeDefined();
      const selectedOffer = discountOffers![0]!;
      const offerTokenValue = selectedOffer.offerTokenAndroid;

      // Create purchase request with the offer token (input field has no suffix)
      const purchaseRequest: RequestPurchaseAndroidProps = {
        skus: [mockProduct.id],
        offerToken: offerTokenValue ?? undefined, // Input field: no suffix
      };

      expect(purchaseRequest.skus).toEqual(['consumable_gems']);
      expect(purchaseRequest.offerToken).toBe('summer_sale_offer_token');
    });
  });

  describe('InstallmentPlanDetailsAndroid', () => {
    it('should have correct structure for subscription installment plans', () => {
      // Installment plan details are available in Google Play Billing Library 7.0+
      const installmentDetails: InstallmentPlanDetailsAndroid = {
        commitmentPaymentsCount: 12,
        subsequentCommitmentPaymentsCount: 12,
      };

      expect(installmentDetails.commitmentPaymentsCount).toBe(12);
      expect(installmentDetails.subsequentCommitmentPaymentsCount).toBe(12);
    });

    it('should support zero subsequentCommitmentPaymentsCount for non-recurring installments', () => {
      // When subsequentCommitmentPaymentsCount is 0, plan reverts to normal after initial commitment
      const nonRecurringInstallment: InstallmentPlanDetailsAndroid = {
        commitmentPaymentsCount: 6,
        subsequentCommitmentPaymentsCount: 0,
      };

      expect(nonRecurringInstallment.commitmentPaymentsCount).toBe(6);
      expect(nonRecurringInstallment.subsequentCommitmentPaymentsCount).toBe(0);
    });
  });

  describe('PendingPurchaseUpdateAndroid', () => {
    it('should have correct structure for pending subscription changes', () => {
      // Pending purchase updates are available in Google Play Billing Library 5.0+
      const pendingUpdate: PendingPurchaseUpdateAndroid = {
        products: ['premium_yearly'],
        purchaseToken: 'pending_token_abc123',
      };

      expect(pendingUpdate.products).toEqual(['premium_yearly']);
      expect(pendingUpdate.purchaseToken).toBe('pending_token_abc123');
    });

    it('should support multiple products in pending update', () => {
      // When upgrading to a bundle or combined subscription
      const bundlePendingUpdate: PendingPurchaseUpdateAndroid = {
        products: ['premium_yearly', 'addon_storage'],
        purchaseToken: 'bundle_pending_token_xyz',
      };

      expect(bundlePendingUpdate.products).toHaveLength(2);
      expect(bundlePendingUpdate.products).toContain('premium_yearly');
      expect(bundlePendingUpdate.products).toContain('addon_storage');
    });
  });

  describe('purchaseOptionIdAndroid on DiscountOffer', () => {
    it('should support purchaseOptionIdAndroid field', () => {
      // purchaseOptionIdAndroid is available in Google Play Billing Library 7.0+
      const discountOfferWithPurchaseOption: DiscountOffer = {
        id: 'limited_offer',
        displayPrice: '$3.99',
        price: 3.99,
        currency: 'USD',
        type: 'one-time',
        offerTokenAndroid: 'offer_token_123',
        purchaseOptionIdAndroid: 'purchase_option_456',
      };

      expect(discountOfferWithPurchaseOption.purchaseOptionIdAndroid).toBe(
        'purchase_option_456',
      );
    });

    it('should allow purchaseOptionIdAndroid to be optional', () => {
      const offerWithoutPurchaseOption: DiscountOffer = {
        displayPrice: '$4.99',
        price: 4.99,
        currency: 'USD',
        type: 'promotional',
        // No purchaseOptionIdAndroid - not all offers have this
      };

      expect(
        offerWithoutPurchaseOption.purchaseOptionIdAndroid,
      ).toBeUndefined();
    });
  });

  describe('PurchaseAndroid with pendingPurchaseUpdateAndroid', () => {
    it('should include pendingPurchaseUpdateAndroid for subscription upgrades', () => {
      // When a subscription upgrade/downgrade is pending
      const purchaseWithPendingUpdate: PurchaseAndroid = {
        id: 'purchase_id_abc',
        productId: 'premium_monthly',
        platform: 'android',
        transactionDate: 1704067200000,
        purchaseToken: 'current_token',
        purchaseState: 'purchased',
        packageNameAndroid: 'com.example.app',
        isAutoRenewing: true,
        quantity: 1,
        store: 'google',
        pendingPurchaseUpdateAndroid: {
          products: ['premium_yearly'],
          purchaseToken: 'pending_upgrade_token',
        },
      };

      expect(
        purchaseWithPendingUpdate.pendingPurchaseUpdateAndroid,
      ).toBeDefined();
      expect(
        purchaseWithPendingUpdate.pendingPurchaseUpdateAndroid?.products,
      ).toEqual(['premium_yearly']);
      expect(
        purchaseWithPendingUpdate.pendingPurchaseUpdateAndroid?.purchaseToken,
      ).toBe('pending_upgrade_token');
    });

    it('should allow pendingPurchaseUpdateAndroid to be undefined for regular purchases', () => {
      const regularPurchase: PurchaseAndroid = {
        id: 'purchase_id_xyz',
        productId: 'consumable_coins',
        platform: 'android',
        transactionDate: 1704067200000,
        purchaseToken: 'regular_token',
        purchaseState: 'purchased',
        packageNameAndroid: 'com.example.app',
        isAutoRenewing: false,
        quantity: 1,
        store: 'google',
        // No pending update - this is a regular purchase
      };

      expect(regularPurchase.pendingPurchaseUpdateAndroid).toBeUndefined();
    });
  });
});

File: libraries/flutter_inapp_purchase/test/standardized_offer_types_test.dart
import 'package:flutter_test/flutter_test.dart';
import 'package:flutter_inapp_purchase/types.dart';

void main() {
  group('DiscountOffer', () {
    test('serializes to JSON with all fields', () {
      const offer = DiscountOffer(
        id: 'discount_001',
        displayPrice: '\$4.99',
        price: 4.99,
        currency: 'USD',
        type: DiscountOfferType.OneTime,
        offerTokenAndroid: 'token_abc',
        offerTagsAndroid: ['discount', 'promo'],
        fullPriceMicrosAndroid: '9990000',
        percentageDiscountAndroid: 50,
        discountAmountMicrosAndroid: '4990000',
        formattedDiscountAmountAndroid: '\$5.00',
      );

      final json = offer.toJson();

      expect(json['id'], 'discount_001');
      expect(json['displayPrice'], '\$4.99');
      expect(json['price'], 4.99);
      expect(json['currency'], 'USD');
      expect(json['type'], 'one-time');
      expect(json['offerTokenAndroid'], 'token_abc');
      expect(json['offerTagsAndroid'], ['discount', 'promo']);
      expect(json['fullPriceMicrosAndroid'], '9990000');
      expect(json['percentageDiscountAndroid'], 50);
    });

    test('deserializes from JSON', () {
      final json = {
        'id': 'discount_002',
        'displayPrice': '\$2.99',
        'price': 2.99,
        'currency': 'USD',
        'type': 'promotional',
        'offerTokenAndroid': 'token_xyz',
      };

      final offer = DiscountOffer.fromJson(json);

      expect(offer.id, 'discount_002');
      expect(offer.displayPrice, '\$2.99');
      expect(offer.price, 2.99);
      expect(offer.type, DiscountOfferType.Promotional);
    });

    test('round-trips through JSON', () {
      const original = DiscountOffer(
        displayPrice: '\$1.99',
        price: 1.99,
        currency: 'EUR',
        type: DiscountOfferType.Introductory,
      );

      final restored = DiscountOffer.fromJson(original.toJson());

      expect(restored.displayPrice, original.displayPrice);
      expect(restored.price, original.price);
      expect(restored.currency, original.currency);
      expect(restored.type, original.type);
    });
  });

  group('SubscriptionOffer', () {
    test('serializes to JSON with cross-platform fields', () {
      const offer = SubscriptionOffer(
        id: 'intro_offer',
        displayPrice: '\$0.99/month',
        price: 0.99,
        currency: 'USD',
        type: DiscountOfferType.Introductory,
        paymentMode: PaymentMode.FreeTrial,
        period: SubscriptionPeriod(
          unit: SubscriptionPeriodUnit.Month,
          value: 1,
        ),
        periodCount: 3,
      );

      final json = offer.toJson();

      expect(json['id'], 'intro_offer');
      expect(json['displayPrice'], '\$0.99/month');
      expect(json['price'], 0.99);
      expect(json['currency'], 'USD');
      expect(json['type'], 'introductory');
      expect(json['paymentMode'], 'free-trial');
      expect(json['period']['unit'], 'month');
      expect(json['period']['value'], 1);
      expect(json['periodCount'], 3);
    });

    test('serializes Android-specific fields', () {
      const phases = PricingPhasesAndroid(
        pricingPhaseList: [
          PricingPhaseAndroid(
            billingCycleCount: 1,
            billingPeriod: 'P1M',
            formattedPrice: 'Free',
            priceAmountMicros: '0',
            priceCurrencyCode: 'USD',
            recurrenceMode: 2,
          ),
        ],
      );

      const offer = SubscriptionOffer(
        id: 'promo_android',
        displayPrice: 'Free',
        price: 0,
        type: DiscountOfferType.Promotional,
        basePlanIdAndroid: 'monthly_plan',
        offerTokenAndroid: 'offer_token_123',
        offerTagsAndroid: ['promo', 'trial'],
        pricingPhasesAndroid: phases,
      );

      final json = offer.toJson();

      expect(json['basePlanIdAndroid'], 'monthly_plan');
      expect(json['offerTokenAndroid'], 'offer_token_123');
      expect(json['offerTagsAndroid'], ['promo', 'trial']);
      expect(json['pricingPhasesAndroid']['pricingPhaseList'], isNotEmpty);
    });

    test('serializes iOS-specific fields', () {
      const offer = SubscriptionOffer(
        id: 'ios_promo',
        displayPrice: '\$4.99',
        price: 4.99,
        type: DiscountOfferType.Promotional,
        keyIdentifierIOS: 'key_123',
        nonceIOS: 'nonce_abc',
        signatureIOS: 'signature_xyz',
        timestampIOS: 1700000000.0,
        localizedPriceIOS: '\$4.99',
        numberOfPeriodsIOS: 6,
      );

      final json = offer.toJson();

      expect(json['keyIdentifierIOS'], 'key_123');
      expect(json['nonceIOS'], 'nonce_abc');
      expect(json['signatureIOS'], 'signature_xyz');
      expect(json['timestampIOS'], 1700000000.0);
      expect(json['localizedPriceIOS'], '\$4.99');
      expect(json['numberOfPeriodsIOS'], 6);
    });

    test('deserializes from JSON with all field variants', () {
      final json = {
        'id': 'offer_full',
        'displayPrice': '\$9.99/year',
        'price': 9.99,
        'currency': 'USD',
        'type': 'introductory',
        'paymentMode': 'pay-as-you-go',
        'period': {'unit': 'year', 'value': 1},
        'periodCount': 1,
        'basePlanIdAndroid': 'yearly_base',
        'offerTokenAndroid': 'token_yearly',
      };

      final offer = SubscriptionOffer.fromJson(json);

      expect(offer.id, 'offer_full');
      expect(offer.price, 9.99);
      expect(offer.type, DiscountOfferType.Introductory);
      expect(offer.paymentMode, PaymentMode.PayAsYouGo);
      expect(offer.period?.unit, SubscriptionPeriodUnit.Year);
      expect(offer.period?.value, 1);
      expect(offer.basePlanIdAndroid, 'yearly_base');
    });

    test('round-trips through JSON', () {
      const original = SubscriptionOffer(
        id: 'roundtrip_offer',
        displayPrice: '\$2.99/week',
        price: 2.99,
        currency: 'GBP',
        type: DiscountOfferType.Promotional,
        paymentMode: PaymentMode.PayUpFront,
        period: SubscriptionPeriod(unit: SubscriptionPeriodUnit.Week, value: 1),
      );

      final restored = SubscriptionOffer.fromJson(original.toJson());

      expect(restored.id, original.id);
      expect(restored.displayPrice, original.displayPrice);
      expect(restored.price, original.price);
      expect(restored.currency, original.currency);
      expect(restored.type, original.type);
      expect(restored.paymentMode, original.paymentMode);
      expect(restored.period?.unit, original.period?.unit);
    });
  });

  group('DiscountOfferType', () {
    test('serializes to kebab-case JSON', () {
      expect(DiscountOfferType.Introductory.toJson(), 'introductory');
      expect(DiscountOfferType.Promotional.toJson(), 'promotional');
      expect(DiscountOfferType.OneTime.toJson(), 'one-time');
    });

    test('deserializes from kebab-case', () {
      expect(
        DiscountOfferType.fromJson('introductory'),
        DiscountOfferType.Introductory,
      );
      expect(
        DiscountOfferType.fromJson('promotional'),
        DiscountOfferType.Promotional,
      );
      expect(DiscountOfferType.fromJson('one-time'), DiscountOfferType.OneTime);
    });
  });

  group('PaymentMode', () {
    test('serializes to kebab-case JSON', () {
      expect(PaymentMode.FreeTrial.toJson(), 'free-trial');
      expect(PaymentMode.PayAsYouGo.toJson(), 'pay-as-you-go');
      expect(PaymentMode.PayUpFront.toJson(), 'pay-up-front');
      expect(PaymentMode.Unknown.toJson(), 'unknown');
    });

    test('deserializes from kebab-case', () {
      expect(PaymentMode.fromJson('free-trial'), PaymentMode.FreeTrial);
      expect(PaymentMode.fromJson('pay-as-you-go'), PaymentMode.PayAsYouGo);
      expect(PaymentMode.fromJson('pay-up-front'), PaymentMode.PayUpFront);
      expect(PaymentMode.fromJson('unknown'), PaymentMode.Unknown);
    });
  });

  group('SubscriptionPeriod', () {
    test('serializes day period', () {
      const period = SubscriptionPeriod(
        unit: SubscriptionPeriodUnit.Day,
        value: 7,
      );

      final json = period.toJson();

      expect(json['unit'], 'day');
      expect(json['value'], 7);
    });

    test('serializes week period', () {
      const period = SubscriptionPeriod(
        unit: SubscriptionPeriodUnit.Week,
        value: 1,
      );

      final json = period.toJson();

      expect(json['unit'], 'week');
      expect(json['value'], 1);
    });

    test('serializes month period', () {
      const period = SubscriptionPeriod(
        unit: SubscriptionPeriodUnit.Month,
        value: 3,
      );

      final json = period.toJson();

      expect(json['unit'], 'month');
      expect(json['value'], 3);
    });

    test('serializes year period', () {
      const period = SubscriptionPeriod(
        unit: SubscriptionPeriodUnit.Year,
        value: 1,
      );

      final json = period.toJson();

      expect(json['unit'], 'year');
      expect(json['value'], 1);
    });

    test('deserializes from JSON', () {
      final json = {'unit': 'month', 'value': 6};

      final period = SubscriptionPeriod.fromJson(json);

      expect(period.unit, SubscriptionPeriodUnit.Month);
      expect(period.value, 6);
    });

    test('round-trips through JSON', () {
      const original = SubscriptionPeriod(
        unit: SubscriptionPeriodUnit.Week,
        value: 2,
      );

      final restored = SubscriptionPeriod.fromJson(original.toJson());

      expect(restored.unit, original.unit);
      expect(restored.value, original.value);
    });
  });

  group('SubscriptionPeriodUnit', () {
    test('serializes to lowercase JSON', () {
      expect(SubscriptionPeriodUnit.Day.toJson(), 'day');
      expect(SubscriptionPeriodUnit.Week.toJson(), 'week');
      expect(SubscriptionPeriodUnit.Month.toJson(), 'month');
      expect(SubscriptionPeriodUnit.Year.toJson(), 'year');
      expect(SubscriptionPeriodUnit.Unknown.toJson(), 'unknown');
    });

    test('deserializes from lowercase', () {
      expect(
        SubscriptionPeriodUnit.fromJson('day'),
        SubscriptionPeriodUnit.Day,
      );
      expect(
        SubscriptionPeriodUnit.fromJson('week'),
        SubscriptionPeriodUnit.Week,
      );
      expect(
        SubscriptionPeriodUnit.fromJson('month'),
        SubscriptionPeriodUnit.Month,
      );
      expect(
        SubscriptionPeriodUnit.fromJson('year'),
        SubscriptionPeriodUnit.Year,
      );
      expect(
        SubscriptionPeriodUnit.fromJson('unknown'),
        SubscriptionPeriodUnit.Unknown,
      );
    });
  });

  group('Product with standardized offers', () {
    test('ProductAndroid includes discountOffers and subscriptionOffers', () {
      const discountOffer = DiscountOffer(
        id: 'discount_1',
        displayPrice: '\$3.99',
        price: 3.99,
        currency: 'USD',
        type: DiscountOfferType.OneTime,
      );

      const product = ProductAndroid(
        currency: 'USD',
        description: 'Test product with discounts',
        displayPrice: '\$9.99',
        id: 'product_with_discounts',
        nameAndroid: 'Test Product',
        platform: IapPlatform.Android,
        price: 9.99,
        title: 'Test Product',
        type: ProductType.InApp,
        discountOffers: [discountOffer],
      );

      final json = product.toJson();

      expect(json['discountOffers'], isNotNull);
      expect(json['discountOffers'], isA<List<dynamic>>());
      expect((json['discountOffers'] as List<dynamic>).length, 1);
      expect((json['discountOffers'] as List).first['type'], 'one-time');
    });

    test('ProductSubscriptionAndroid includes subscriptionOffers', () {
      const subscriptionOffer = SubscriptionOffer(
        id: 'intro_offer',
        displayPrice: '\$0.99/month',
        price: 0.99,
        type: DiscountOfferType.Introductory,
        paymentMode: PaymentMode.FreeTrial,
      );

      const phases = PricingPhasesAndroid(
        pricingPhaseList: [
          PricingPhaseAndroid(
            billingCycleCount: 1,
            billingPeriod: 'P1M',
            formattedPrice: '\$4.99',
            priceAmountMicros: '4990000',
            priceCurrencyCode: 'USD',
            recurrenceMode: 1,
          ),
        ],
      );

      const legacyOffer = ProductSubscriptionAndroidOfferDetails(
        basePlanId: 'monthly',
        offerToken: 'token_123',
        offerTags: ['intro'],
        pricingPhases: phases,
      );

      const subscription = ProductSubscriptionAndroid(
        currency: 'USD',
        description: 'Monthly subscription',
        displayPrice: '\$4.99/month',
        id: 'monthly_sub',
        nameAndroid: 'Monthly Sub',
        platform: IapPlatform.Android,
        price: 4.99,
        subscriptionOfferDetailsAndroid: [legacyOffer],
        subscriptionOffers: [subscriptionOffer],
        title: 'Monthly Subscription',
        type: ProductType.Subs,
      );

      final json = subscription.toJson();

      // Has both legacy and new fields
      expect(json['subscriptionOfferDetailsAndroid'], isNotNull);
      expect(json['subscriptionOffers'], isNotNull);

      // New standardized offer
      expect(
        (json['subscriptionOffers'] as List).first['paymentMode'],
        'free-trial',
      );
    });
  });
}
